prometheus/example_timer_batch_test.go

```diff
@@ -0,0 +1,48 @@
+// Copyright 2014 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prometheus_test
+
+import (
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var (
+	// For periodic or batch processes that run much less frequently than
+	// the Prometheus scrape, it makes sense to use a Gauge to observe the
+	// duration.
+	// See https://prometheus.io/docs/practices/instrumentation/#batch-jobs
+	batchDuration = prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "example_batch_duration_seconds",
+		Help: "Duration of the last batch run.",
+	})
+)
+
+func performBatch() {
+	// The Set method of the Gauge is used to observe the duration.
+	timer := prometheus.NewTimer(prometheus.ObserverFunc(batchDuration.Set))
+	defer timer.ObserveDuration()
+
+	// Actually perform the batch of work.
+	// ...
+}
+
+func ExampleTimer_batch() {
```

Comment:
I can change it into a PGW example. However, we are using a Gauge for checkpoint duration in Prometheus itself, and that always appeared very useful to me.

Just yesterday, I ran into a user at SC who used a histogram in a case similar to Prometheus checkpoints, and in that situation, a Gauge would have been much easier to reason with, too.

If something should be timed that happens much less frequent than a scrape, a Gauge as timer should be considered. I would like to see that in the best practices (where it currecntly says, there are only two valid use cases for a Gauge for timing, non-trivial Collectors and "true" batch jobs).

However, to not open up another front of the discussion, I'll change the example to PGW, and we can discuss the Gauge for timing issue later..